### PR TITLE
Anonymize feedback and add unit tests for the method added to the UtilModel (Fixed #1615)

### LINKED ISSUE #1615: Anonymize feedback section
**Actual Behaviour**

The feedback section currently shows complete email addresses. 

**Expected Behaviour**

Anonymize the feedback and only show emails until the @ sign, e.g. "abc@..".

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by arundhati24**
Fixes #1615 

Changes: Now the complete email is not shown. Rather, an email is truncated after '@' symbol.

**Screenshot**

truncateemail

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/skills/feedback/adapters/recycleradapters/AllReviewsAdapter.kt [hunk 1]
```diff
@@ -39,7 +39,10 @@ class AllReviewsAdapter(val context: Context, val feedbackList: List<Feedback>?)
                 if (feedbackList[position].email != null &&
                         !TextUtils.isEmpty(feedbackList[position].email)) {
                     holder.initials.text = feedbackList[position].email?.substring(0, 2)
-                    holder.feedbackEmail?.text = feedbackList[position].email
+                    val truncateAt = feedbackList.get(position).email?.indexOf('@')
+                    if (truncateAt != null && truncateAt is Int) {
+                        holder.feedbackEmail?.text = feedbackList.get(position).email?.substring(0, truncateAt.plus(1)) + " ..."
+                    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/feedback/adapters/recycleradapters/AllReviewsAdapter.kt]: Add the method in Utils as this is a helper method reused in both files

Duplication is bad

Also, add tests

**Comment by arundhati24** [comment]
@iamareebjamal I made the changes and also added test for the added method. Please have a look.

File: app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt [hunk 2]
```diff
@@ -17,6 +17,15 @@ import retrofit2.Response
...
 
+    companion object {
+        fun truncateEmailAtEnd(email: String?): String? {
+            val truncateAt = email?.indexOf('@')
+            if (truncateAt is Int) {
+                return (email.substring(0, truncateAt.plus(1)) + " ...")
+            }
+            return null
+        }
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt]: Should be in Utils object

File: app/src/test/java/data/UtilModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Created by arundhati24 on 03/08/2018.
+ */
+class UtilModelTest {
+
+    @Test
+    fun testTruncateEmailAtEnd() {
+        val actualEmail = UtilModel.truncateEmailAtEnd("testuser@example.com")
+        val expectedEmail = "testuser@ ..."
+        assert(actualEmail?.trim().equals(expectedEmail))
```
Comment (iamareebjamal) [inline app/src/test/java/data/UtilModelTest.kt]: Don't trim and use `assertEquals`

Comment (iamareebjamal) [reply app/src/test/java/data/UtilModelTest.kt]: Add 2 3 more tests for edge cases where the test might fail

File: app/src/test/java/helper/UtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.Test
+
+/**
+ *
+ * Created by arundhati24 on 03/08/2018.
+ */
+class UtilsTest {
+    @Test
+    fun testTruncateEmailAtEnd() {
+        assertEquals(true, Utils.truncateEmailAtEnd("testuser@example.com").equals("testuser@ ..."))
```
Comment (iamareebjamal) [inline app/src/test/java/helper/UtilsTest.kt]: Wrong use of assetEquals

File: app/src/test/java/helper/UtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package helper
```
Comment (iamareebjamal) [inline app/src/test/java/helper/UtilsTest.kt]: Use the same package as the file to be tested

**Comment by iamareebjamal** [comment]
@arundhati24 In how many days will this be fixed?

**Comment by arundhati24** [comment]
@iamareebjamal It will be fixed today itself.

File: app/src/test/java/helper/UtilsTest.kt (hunk 4 above)
Comment (arundhati24) [reply app/src/test/java/helper/UtilsTest.kt]: @iamareebjamal assertEquals() takes two params expected value and actual value, so is it because of the passed params?

Comment (iamareebjamal) [reply app/src/test/java/helper/UtilsTest.kt]: `truncateEmailAtEnd` returns String and not boolean. So expected param should be?

**Comment by arundhati24** [comment]
@iamareebjamal Please review this PR. (Codacy only shows two documentation issues, which need to be ignored I think).

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import org.junit.Test
+
+/**
+ *
+ * Created by arundhati24 on 03/08/2018.
+ */
+class UtilsTest {
+    @Test
+    fun testTruncateEmailAtEnd() {
+        assertEquals("testuser@ ...", Utils.truncateEmailAtEnd("testuser@example.com"))
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Add more emails which may produce errors. Like emails with period and special symbols. Goad of tests is to break the code as much it is to test if it is working correctly

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        assertEquals(null, Utils.truncateEmailAtEnd("testuser"))
+        assertEquals(null, Utils.truncateEmailAtEnd(""))
+        assertEquals(null, Utils.truncateEmailAtEnd(" "))
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Add test for null

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt [hunk 8]
```diff
@@ -58,4 +57,13 @@ object Utils {
         return result.toString()
     }
 
+    fun truncateEmailAtEnd(email: String?): String? {
+        if (!email.isNullOrEmpty()) {
+            val truncateAt = email?.indexOf('@')
+            if (truncateAt is Int && truncateAt != -1) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: `truncateAt is Int` is not required?

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt (hunk 6 above)
Comment (iamareebjamal) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: What about emails with multiple `@` symbols?

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt [hunk 9]
```diff
@@ -58,4 +57,13 @@ object Utils {
...
+                return (email.substring(0, truncateAt.plus(1)) + " ...")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: Extra paranthesis

File: app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt (hunk 8 above)
Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: As email is nullable, ```truncateAt``` is being called upon a nullable String object. Later when I have used ```truncateAt.plus(1)``` for that we strictly require ```Int``` but ```truncateAt``` is of type ```Int?```
Hence, I have used ```truncateAt is Int``` when the email is not null or empty.
Could this be improved?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/helper/Utils.kt]: `truncateAt != null and ...`

File: app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt (hunk 6 above)
Comment (arundhati24) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: @iamareebjamal As this type of email is invalid, hence null should be returned. To check for this, I tried checking for the validity of the email in the ```truncateEmailAtEnd()``` using the ```CredentialsHelper.kt``` class which in turn uses the TextUtils class to check for the validity of the email. But, the tests fail with the error ```java.lang.RuntimeException: Method isEmpty in android.text.TextUtils not mocked. See http://g.co/androidstudio/not-mocked for details```

On checking the link, I found this :

screen shot 2018-08-10 at 8 04 51 pm

On adding the above code snippet to build.gradle and then running the tests I received an NPE (though I had checked for nullability of email and then used !!)

```
fun truncateEmailAtEnd(email: String?): String? {
        if (!email.isNullOrEmpty()) {
            if (CredentialHelper.isEmailValid(email!!)) {
                val truncateAt = email.indexOf('@')
                if (truncateAt is Int && truncateAt != -1) {
                    return email.substring(0, truncateAt.plus(1)) + " ..."
                }
            }
        }
        return null
    }
```

Comment (iamareebjamal) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: You cannot use android classes in java

Comment (arundhati24) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: Would you please suggest how should I account for the cases of multiple '@' or '@' followed by any random symbols?

Comment (iamareebjamal) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: I think that it should simply be truncated after a threshold amount of characters, not according to the position of `@` symbol

Comment (arundhati24) [reply app/src/test/java/org/fossasia/susi/ai/helper/UtilsTest.kt]: @iamareebjamal I feel what you have mentioned is true. But, as this is how this has been done on Skills CMS and iOS to anonymise the identity of the user, I need to implement it in the same way in the Android app.
Hence, please suggest what I should do in order to add tests for such a case (although an invalid email is not expected as the response is coming from the server).
